src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionBase.cs

```diff
@@ -0,0 +1,490 @@
+                    nameof(Client.Constants.Documents.SubscriptionChangeVectorSpecialStates.DoNotChange),
+                    NonExistentBatch, docsToResend: null);
+        }
+
+        internal async Task<SubscriptionConnectionClientMessage> GetReplyFromClientAsync()
+        {
+            try
+            {
+                using (TcpConnection.ContextPool.AllocateOperationContext(out JsonOperationContext context))
+                using (var blittable = await context.ParseToMemoryAsync(
+                           TcpConnection.Stream,
+                           "Reply from subscription client",
+                           BlittableJsonDocumentBuilder.UsageMode.None,
+                           _copiedBuffer.Buffer))
+                {
+                    TcpConnection._lastEtagReceived = -1;
+                    TcpConnection.RegisterBytesReceived(blittable.Size);
+                    return JsonDeserializationServer.SubscriptionConnectionClientMessage(blittable);
+                }
+            }
+            catch (EndOfStreamException e)
+            {
+                throw new SubscriptionConnectionDownException("No reply from the subscription client.", e);
+            }
+            catch (IOException)
+            {
+                if (_isDisposed == false)
+                    throw;
+
+                return new SubscriptionConnectionClientMessage
+                {
+                    ChangeVector = null,
+                    Type = SubscriptionConnectionClientMessage.MessageType.DisposedNotification
+                };
+            }
+            catch (ObjectDisposedException)
+            {
+                return new SubscriptionConnectionClientMessage
+                {
+                    ChangeVector = null,
+                    Type = SubscriptionConnectionClientMessage.MessageType.DisposedNotification
+                };
+            }
+        }
+
+
+        public void Dispose()
+        {
+            using (_copiedBuffer.ReleaseBuffer)
+            {
+                try
+                {
+                    CancellationTokenSource.Cancel();
+                }
+                catch
+                {
+                    // ignored
+                }
+
+                try
+                {
+                    _tcpConnectionDisposable?.Dispose();
+                }
+                catch
+                {
+                    // ignored
+                }
+
+                try
+                {
+                    TcpConnection.Dispose();
+                }
+                catch
+                {
+                    // ignored
+                }
+
+
+                try
+                {
+                    CancellationTokenSource.Dispose();
+                }
+                catch
+                {
+                    // ignored
+                }
+
+                RecentSubscriptionStatuses?.Clear();
+            }
+        }
+    }
+
+    public class ShardData
```

Comment:
not related to this file? why it is here?

Reply:
moved inside the class